modules/linear_boltzmann_solvers/lbs_solver/volumetric_source/volumetric_source.cc

```diff
@@ -0,0 +1,125 @@
+// SPDX-FileCopyrightText: 2024 The OpenSn Authors <https://open-sn.github.io/opensn/>
+// SPDX-License-Identifier: MIT
+
+#include "modules/linear_boltzmann_solvers/lbs_solver/volumetric_source/volumetric_source.h"
+#include "modules/linear_boltzmann_solvers/lbs_solver/lbs_solver.h"
+#include "framework/mesh/cell/cell.h"
+#include "framework/mesh/mesh_continuum/mesh_continuum.h"
+#include "framework/mesh/logical_volume/logical_volume.h"
+#include "framework/math/functions/spatial_material_function.h"
+#include "framework/runtime.h"
+#include "framework/object_factory.h"
+
+namespace opensn::lbs
+{
+
+OpenSnRegisterObjectInNamespace(lbs, VolumetricSource);
+
+InputParameters
+VolumetricSource::GetInputParameters()
+{
+  InputParameters params = Object::GetInputParameters();
+
+  params.SetGeneralDescription("General implementation of an arbitrary multi-group "
+                               "distributed source. A distributed source is given by a "
+                               "logical volume and a Lua function handle. Currently, only "
+                               "isotropic distributed sources are allowed.");
+  params.SetClassName("Distributed Source");
+  params.SetDocGroup("LBSUtilities");
+
+  params.AddOptionalParameterArray(
+    "block_ids",
+    std::vector<int>(),
+    "An array of block IDs the volumetric source is present within.");
+  params.AddOptionalParameterArray("group_strength",
+                                   std::vector<double>(),
+                                   "An array of multi-group source strength values. Note that this "
+                                   "is only used when a function is not provided.");
+  params.AddOptionalParameter(
+    "logical_volume_handle",
+    SIZE_T_INVALID,
+    "Handle to the logical volume the volumetric source is defined within.");
+  params.AddOptionalParameter(
+    "function_handle",
+    SIZE_T_INVALID,
+    "Handle to a ResponseFunction object to be used to define the source.");
+
+  return params;
+}
+
+VolumetricSource::VolumetricSource(const InputParameters& params)
+  : block_ids_(params.GetParamVectorValue<int>("block_ids")),
+    logvol_(params.ParametersAtAssignment().Has("logical_volume_handle")
+              ? GetStackItemPtrAsType<LogicalVolume>(
+                  object_stack, params.GetParamValue<size_t>("logical_volume_handle"))
+              : nullptr),
+    strength_(params.GetParamVectorValue<double>("group_strength")),
+    function_(params.ParametersAtAssignment().Has("function_handle")
+                ? GetStackItemPtrAsType<SpatialMaterialFunction>(
```

Comment:
also the name SpatialMaterialFunction feels weird now. When doing the adjoint, the distributed/volumetric source was often associated with a material (the "detector cross section"). But now that distributed sources can be used in forward mode, I do not think that associating a source with a material is what we wanted to keep. It is now a bit awkward to me to read SpatialMaterialFunction here now.

Reply:
> also the name SpatialMaterialFunction feels weird now. 

`SpatialMaterialFunction` just describes the arguments to an `Evaluate` call in the function system. It takes a cell, a node, a material(block) ID, and a number of components. The `SpatialMaterialFunction` has been an attribute of the `Distributed(Volumetric)Source` since the function system was put into the code awhile back. 

> When doing the adjoint, the distributed/volumetric source was often associated with a material (the "detector cross section"). But now that distributed sources can be used in forward mode, I do not think that associating a source with a material is what we wanted to keep. 

The volumetric source in its previous state was applied to cells within a logical volume. The user could pass a function that performs operations based on the arguments stated above that could 1) provide spatial and group-dependence within the logical volume, and 2) filter for particular material/block IDs. If was not function is passed, then the default behavior of a unit source in each group within the logical volume was used. This is where the tie to materials came into play. It has nothing to do with the `Material` object, like `IsotropicMultiGroupSource` does.

I'm not sure exactly what you mean by "I do not think that associating a source with a material is what we wanted to keep" since the `VolumetricSource` is associated with block IDs from the mesh and/or logical volumes not `Material` objects. Anything other than this would be a departure from how sources are currently specified in the code. As I state in the description, this PR is meant to move the functionality of `IsotropicMultiGroupSource` out of something that lives on `Material`, not to redo the source specification system.

> It is now a bit awkward to me to read SpatialMaterialFunction here now.

Do the above comments make this less awkward? Otherwise, I'm not sure how to address this.

Reply:
We are trying to undo the connection of source->material object (big scheme of things). In that process, I do not think we should have the connection between a volumetric source and either an LV or a set of blockIDs (using the future name here) be called SpatialMaterialFunction. The source is in a volume, which is identified by an LV or block ID. It is the Material portion of the name that bugs me.

Reply:
I had typed up a response and thought it submitted but it was marked as "pending" in a review I must have accidentally started. I'm not entirely sure what the particular change you are asking for is. I understand your concern, but how would you like me to address it?  Change the name of `SpatialMaterialFunction`, or implement a new function in the function system that only takes a spatial coordinate called something like `SpatialFunction` (along with a `LuaSpatialFunction` counterpart)? If the prior, what would you suggest changing the name to that contains the same information? `SpatialMaterialIDFunction`? `SpatialBlock(ID)Function`?

Reply:
Request (I think). It is the Material portion of the function name that feels strange to me. The source will be turned on based on a LV or a set of (soon to be named) block IDs. So the Material portion of the name is what I was referring to

Reply:
We discussed this last Wednesday and I will be modifying the function system class `SpatialMaterialFunction` to remove the material ID function input and rename it to something like `SpatialFunction` or `VectorSpatialFunction` to mirror `ScalarSpatialFunction` which outputs a scalar value.

I should reiterate, however, that `SpatialMaterialFunction` is not something defined in this PR or within the `VolumetricSource` class. It is a generic function in the function system. Its use here was due to the fact that it was the only available function in the function system that had the desired input/output behavior. While I don't think that we should be renaming generic objects in `framework` based on individual use cases, because there are no other use cases for `SpatialMaterialFunction` in the code and we do not have a more general function the returns a vector from  a spatial coordinate argument, this modification is worthwhile here.

Reply:
I believe I knew that `SpatialMaterialFunction` was not new to the PR. It was just that the name did not make sense (or no longer made sense).